sound/soc/codecs/rt711-sdca-sdw.c

```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// rt711-sdw-sdca.c -- rt711 SDCA ALSA SoC audio driver
+//
+// Copyright(c) 2020 Realtek Semiconductor Corp.
+//
+//
+
+#include <linux/delay.h>
+#include <linux/device.h>
+#include <linux/mod_devicetable.h>
+#include <linux/soundwire/sdw_registers.h>
+#include <linux/module.h>
+
+#include "rt711-sdca.h"
+#include "rt711-sdca-sdw.h"
+
+static bool rt711_sdca_readable_register(struct device *dev, unsigned int reg)
+{
+	switch (reg) {
+	case SDW_DP0_INT:
+	case SDW_SCP_SDCA_INT1 ... SDW_SCP_SDCA_INTMASK4:
+	case 0x201a ... 0x2027:
+	case 0x2029 ... 0x202a:
+	case 0x202d ... 0x2034:
+	case 0x2200 ... 0x2204:
+	case 0x2206 ... 0x2212:
+	case 0x2220 ... 0x2223:
+	case 0x2230 ... 0x2239:
+	case 0x2f01 ... 0x2f0f:
+	case 0x2f30 ... 0x2f36:
+	case 0x2f50 ... 0x2f5a:
+	case 0x2f60:
+	case 0x3200 ... 0x3212:
+	case 0x2000000 ... 0x20000ff:
+	case 0x2002000 ... 0x20020ff:
+	case 0x5600000 ... 0x56000ff:
+	case 0x5602000 ... 0x56020ff:
+	case 0x5700000 ... 0x57000ff:
+	case 0x5702000 ... 0x57020ff:
+	case 0x5800000 ... 0x58000ff:
+	case 0x5802000 ... 0x58020ff:
+	case 0x5900000 ... 0x59000ff:
+	case 0x5902000 ... 0x59020ff:
+	case 0x5b00000 ... 0x5b000ff:
+	case 0x5b02000 ... 0x5b020ff:
+	case 0x5f00000 ... 0x5f000ff:
+	case 0x5f02000 ... 0x5f020ff:
+	case 0x6100000 ... 0x61000ff:
+	case 0x6102000 ... 0x61020ff:
+	case 0x40600488 ... 0x40600490:
+	case 0x40c80080 ... 0x40c80098:
+	case 0x44030000 ... 0x44030017:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool rt711_sdca_volatile_register(struct device *dev, unsigned int reg)
+{
+	switch (reg) {
+	case SDW_DP0_INT:
+	case SDW_SCP_SDCA_INT1 ... SDW_SCP_SDCA_INT4:
+	case 0x201b:
+	case 0x201c:
+	case 0x201d:
+	case 0x201f:
+	case 0x2021:
+	case 0x2023:
+	case 0x2230:
+	case 0x202d ... 0x202f: /* BRA */
+	case 0x2200 ... 0x2212: /* i2c debug */
+	case RT711_RC_CAL_STATUS:
+	case 0x200001a:
+	case 0x200201a:
+	case 0x2000046:
+	case 0x2002046:
+	case 0x2000080:
+	case 0x2002080:
+	case 0x2000081:
+	case 0x2002081:
+	case 0x2000083:
+	case 0x2002083:
+	case 0x5800000:
+	case 0x5802000:
+	case 0x5800001:
+	case 0x5802001:
+	case 0x5f00001:
+	case 0x5f02001:
+	case 0x40600490:
+	case 0x40c80080 ... 0x40c80098:
+	case 0x44030000 ... 0x44030017:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static const struct regmap_config rt711_sdca_regmap = {
+	.reg_bits = 32,
+	.val_bits = 8,
+	.readable_reg = rt711_sdca_readable_register,
+	.volatile_reg = rt711_sdca_volatile_register,
+	.max_register = 0x44ffffff,
+	.reg_defaults = rt711_sdca_reg_defaults,
+	.num_reg_defaults = ARRAY_SIZE(rt711_sdca_reg_defaults),
+	.cache_type = REGCACHE_RBTREE,
+	.use_single_read = true,
+	.use_single_write = true,
+};
+
+static const struct regmap_config rt711_sdca_mbq_regmap = {
+	.name = "sdw-mbq",
+	.reg_bits = 32,
+	.val_bits = 16,
+	.readable_reg = rt711_sdca_readable_register,
+	.volatile_reg = rt711_sdca_volatile_register,
```

Comment:
I would have put separate functions for MBQ registers, that would make a clean cut between the two regular regmap and MBQ-based ones.

Reply:
OK, I will create separate functions for MBA registers.